source/develop/release-management/features/virt/engine-dedicated-cpus.md

```diff
@@ -0,0 +1,173 @@
+---
+Title: Engine Dedicated CPUs
+Category: feature
+Authors: sabusale
+---
+
+# Engine Dedicated CPUs
+
+## Summary
+
+Once choosing dedicated CPU policy, The guest vCPUs will be strictly pinned to a set of host pCPUs
+(similarly to static CPU pinning). The set of pCPUs will be chosen to match the required guest CPU topology.
+
+Engine sets the pinning automatically when starting the VM as opposed to the manual pinning that is
+done by the user and is static.
+
+## Owner
+
+-   Name: Saif Abu Saleh (sabusale)
+-   Email: sabusale@redhat.com
+
+## Responsibilities
+
+VDSM:
+
+- Makes sure that dedicated VMs vCPUs stay exclusively pinned to the given pCPU
+
+Engine:
+- Holds cached available cpus based on host capabilities and pinned CPUs
+- Schedule VM to run on best satisfying host based on the above cached data structure
+
+
+
+## Backend
+
+1. Add the new policies to CPU Pinning Policy enum
+2. Save host capabilities CPU topology in vds_dynamic
+3. Hold hostToAvailableCpu data structure:
+    it will be on this form `Map<Integer,Map<Integer,Map<Integer,List<String>>>>`
+   
+    Host id -> Map of Socket id to Map of Core id to list of cpus
```

Comment:
if we store it in a per-host entity like VdsManager then we can save that first mapping

Reply:
I would also prefer to use java classes instead of this Map structure - for better readability + it would allow to add also some functionality to the classes. E.g. VdsSocket -> VdsCore -> VdsCpu. Possible methods could be VdsCore:hasAvailableCpus, VdsCore:getAvailableCpus, VdsCpu:isAvailable, maybe even VdsCpu:getVm (to return VM that this CPU is dedicated to).

Reply:
Done